File: digital_ebook_library/lib/models/ebook.dart
class Ebook {
  final int id;
  final String title;
  final String? author;
  final String? fileType;
  final int? fileSize;
  final String? fileName;
  final DateTime createdAt;
  final String? coverUrl;
  final String? downloadUrl;

  Ebook({
    required this.id,
    required this.title,
    this.author,
    this.fileType,
    this.fileSize,
    this.fileName,
    required this.createdAt,
    this.coverUrl,
    this.downloadUrl,
  });

  factory Ebook.fromJson(Map<String, dynamic> json) {
    return Ebook(
      id: json['id'],
      title: json['title'] ?? 'Untitled',
      author: json['author'],
      fileType: json['file_type'],
      fileSize: json['file_size'],
      fileName: json['file_name'],
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      coverUrl: json['cover_url'],
      downloadUrl: json['download_url'],
    );
  }

  String get fileSizeLabel {
    if (fileSize == null) return '';
    final kb = fileSize! / 1024;
    if (kb < 1024) return '${kb.toStringAsFixed(0)} KB';
    return '${(kb / 1024).toStringAsFixed(1)} MB';
  }

  bool get isPdf => (fileType ?? '').contains('pdf');
}

File: ebook_library_backend/app/models/ebook.rb
class Ebook < ApplicationRecord
  has_one_attached :file
  has_one_attached :cover

  MAX_FILE_SIZE = 50.megabytes
  ALLOWED_TYPES = %w[application/pdf application/epub+zip].freeze

  validates :title, presence: true
  validates :file, presence: true
  validate :acceptable_file

  before_validation :set_file_metadata

  scope :search, ->(query) {
    return all if query.blank?

    where("title LIKE :q OR author LIKE :q OR file_name LIKE :q", q: "%#{query}%")
  }

  def as_json_summary
    {
      id: id,
      title: title,
      author: author,
      file_type: file_type,
      file_size: file_size,
      file_name: file_name,
      created_at: created_at,
      cover_url: cover.attached? ? Rails.application.routes.url_helpers.rails_blob_path(cover, only_path: true) : nil,
      download_url: file.attached? ? "/api/ebooks/#{id}/download" : nil
    }
  end

  private

  def set_file_metadata
    return unless file.attached?

    self.file_type ||= file.content_type
    self.file_size ||= file.byte_size
    self.file_name ||= file.filename.to_s
  end

  def acceptable_file
    return unless file.attached?

    unless ALLOWED_TYPES.include?(file.content_type)
      errors.add(:file, "must be a PDF or EPUB")
    end

    if file.byte_size > MAX_FILE_SIZE
      errors.add(:file, "is too large (max #{MAX_FILE_SIZE / 1.megabyte}MB)")
    end
  end
end